otherlibs/runtime_events/runtime_events.ml

```diff
@@ -258,14 +258,35 @@ module User = struct
        and we never free those buffers. *)
     let create_buffer () = Bytes.create 1024 in
     let write_buffer_cache = Domain.DLS.new_key (fun () -> ref []) in
+    let pop_or_create buffers =
+      (* intended to be thread-safe *)
+      (* begin atomic *)
+      match !buffers with
+      | [] ->
+          (* end atomic *)
+          create_buffer ()
+      | b::bs ->
+          buffers := bs;
+          (* end atomic *)
+          b
+    in
+    let rec push buffers buf =
+      (* intended to be thread-safe *)
+      let cur_buffers = !buffers in
+      let cons = buf :: cur_buffers in
+      (* retry if !buffers changed under our feet: *)
+      (* begin atomic *)
+      if !buffers == cur_buffers then
```

Comment:
I fear that the compiler might perform CSE, making this safeguard ineffective.

Reply:
I stand corrected, it looks like allocations cause equations over mutable loads to be dropped.

Reply:
What you should worry about here is that the compiler is 100% allowed to move the allocation into the branch (after the test).
You can use `Sys.opaque_identity` around the allocation to prevent that, but in the end there is no specification that would prevent the compiler from inserting a poll between the equality check and the assignment so you can't be provably safe.

Reply:
You know better than me about this, but it is not easy to work with the fact that there is no documentation of what "the compiler is allowed to do" -- I agree with you that having more precise semantics for this would be useful. Is this allowed in Flambda land? I would have thought that `!buffers` (in the test `!buffers == cur_buffers`) is considered as having coeffects, and that `buf :: cur_buffers` is considered as having generative effects, and that the two thus cannot be permuted -- but I lost myself trying to check the implementation for this.

Reply:
> Is this allowed in Flambda land?

Technically yes, our semantics would allow this. But I don't think we have implemented any transformations that would make use of that in Flambda 1.

> ... is considered as having generative effects, and that the two thus cannot be permuted

Generative effects are a different category of effects. The only generative effects we track are memory allocations, and we do not track co-effects in this category, so they are allowed to be permuted with everything. Other examples of operations that could have generative effects only could be random number generation, if that was a language builtin, or monotonic counters such as `caml_fresh_oo_id`.
If we wanted to accurately model the runtime execution we could give memory co-effects to everything that interacts with the GC (polls and C calls, typically), but that would also mean that we wouldn't be allowed to remove any allocation if there might be a poll later in the program (just like you cannot remove an assignment to a reference if there might be a read from it later). So not very useful for an optimiser.

Reply:
I think it would be interesting to treat an allocation as the combination of a poll point and a generative effect, and then discuss the semantic guarantees we want for poll points.

Moving a memory read across a poll point is not valid in my book, unless we know for sure that the memory is local to the current thread and thus cannot be mutated by another thread, domain or signal handler. (So for example reads in local mutable references would move around freely.) 

On the other hand, reordering poll points should be fine, and I think that we could allow "fusing" poll points, that is dropping one if we know that there is another close by. (Because allocation poll points are tied to the allocation itself, we can only drop the poll point if we decide to also discard the allocated value.)

This would correspond to a thread and memory model where there may be less interleaving than suggested by a naive reading of the code (... if it is aware of the implicit placement of poll points), but we can reason about reads happening before or after the context switches. This is the reasoning in CSEgen as @OlivierNicole pointed out, do we see horrible outcomes if the rest of the compiler followed it?

https://github.com/ocaml/ocaml/blob/a682d5196ade94cbb04105e48ba2341d9e71a7aa/asmcomp/CSEgen.ml#L300-L303

Reply:
In the short term and in the context of this issue, @lthls, I would be curious to know if you believe that the code as currently written is fine, or if you would recommend doing things differently.

Reply:
My opinion is that this code should use atomic cells instead of references if consistency is required. If not (i.e. if we can deal with dropping a few bits of data from time to time), then `let push buffers buf = buffers := buf :: !buffers` will work.

Reply:
> My opinion is that this code should use atomic cells instead of references

Mph, I don't want to do this because the point is precisely to avoid any synchronization bottleneck between runtime_event users. I use Domain.DLS to avoid synchronization, I don't want to use Atomic for single-domain multi-threaded programming.

I reformulated the code using an explicit compare_and_set operation that is annotated `@poll error` (this implies `[@inline never]`), does that approach look correct to you?

Reply:
It looks more robust (although I don't consider the `[@poll error]` mechanism as foolproof), but I wouldn't be surprised if it ends up more expensive than regular atomic operations.

Reply:
I'll take "expansive like a direct function call" over "complex stuff that might create weird bottlenecks on relaxed machines and work fine everywhere else" any day.